src/targets/gpu/kernels/include/migraphx/kernels/index.hpp

```diff
@@ -36,44 +36,33 @@ namespace migraphx {
 #if defined(MIGRAPHX_NGLOBAL) && defined(MIGRAPHX_NLOCAL)
 #define MIGRAPHX_NGROUP ((MIGRAPHX_NGLOBAL + MIGRAPHX_NLOCAL - 1) / MIGRAPHX_NLOCAL)
 #endif
+inline __device__ __attribute__((const)) index_int compute_max_local_size()
+{
+#ifdef MIGRAPHX_NLOCAL
+    return MIGRAPHX_NLOCAL;
+#else
+    MIGRAPHX_ASSERT(compute_global_size() % blockDim.x == 0);
+    return blockDim.x;
+#endif
+}
 
 inline __device__ __attribute__((const)) index_int compute_global_size()
 {
 #ifdef MIGRAPHX_NGLOBAL
     return MIGRAPHX_NGLOBAL;
 #else
-    // This actualy works even when global is not divisible by local size.
-    // This doesnt actually do a multiplicatiosn. Instead it calls a device
-    // function to get the global size, which is why it works.
-    return blockDim.x * gridDim.x;  // NOLINT
+    return compute_max_local_size() * gridDim.x; // NOLINT
 #endif
 }
 
-// We cant just use blockDim.x to get the local size since its broken on hip
-// when global is not divisible by local size. In this case, we calulate the
-// size for the last group.
 inline __device__ __attribute__((const)) index_int compute_local_size()
 {
 #ifdef MIGRAPHX_NLOCAL
     const auto nlocal = MIGRAPHX_NLOCAL;
 #else
-    const auto nlocal = blockDim.x; // NOLINT
-#endif
-#ifdef MIGRAPHX_NGROUP
-    const auto ngroup = MIGRAPHX_NGROUP;
-#else
-    const auto ngroup = gridDim.x;  // NOLINT
+    const auto nlocal = blockDim.x;              // NOLINT
 #endif
-    const auto group_id = blockIdx.x; // NOLINT
-    const auto nglobal  = compute_global_size();
-    if(group_id == ngroup - 1)
-    {
-        return 1 + (nglobal - 1) % nlocal;
-    }
-    else
-    {
-        return nlocal; // NOLINT
-    }
+    return nlocal; // NOLINT
```

Comment:
This is wrong as well. This will return the max size when `MIGRAPHX_NLOCAL` is defined. It should do this instead:

```cpp
inline __device__ __attribute__((const)) index_int compute_local_size()
{
#ifdef MIGRAPHX_HAS_CONST_LOCAL
    return MIGRAPHX_NLOCAL;
#else
    return blockDim.x;
#endif
}
```

And instead of relying on `blockDim.x` which seems to change behavior with non-uniform blocks, we could call `__ockl_get_local_size(0)` directly to get the correct size.